### DISCUSSION TITLE
Need some help migrating from ethers

### DISCUSSION THREAD

**Discussion by onchainguy-btc** [wevm/viem#605] [Question]
I've got the following code that is based on ethers. I checked the migration guide but couldn't really find anything useful to refactor this code. Any help would be appreciated:

```typescript
import { useEffect, useMemo, useState } from 'react';
import { CallOverrides, Contract, ethers } from 'ethers';
import { useSigner } from 'wagmi';
import { useChain } from '../use-chain';
import { getAddressByGroup } from '@polkamon/util-essential';
import { ContractArgs, UseContractProps } from './use-contract.types';

async function buildContractCall<CustomContractArgs>(
  contract: Contract,
  functionName: string,
  args?: (ContractArgs | CustomContractArgs)[],
  overrides?: CallOverrides
): Promise<string | undefined> {
  const { data } = await contract.populateTransaction[functionName](...(args || []), overrides);
  return data;
}

// this function goes through ethers instead of wagmi, y? Example below:
// Transacting via credit card needs the raw contract input data. That can be returned from wagmi, too.
// The problem is that wagmi will simulate the contract call and crash if the user has not enough balance.
// Well, when the user wants to pay via credit card then probably actually because they have no balance.
// This means we need to build up a raw call instead.
export function useContractRaw<CustomContractArgs>({
  addresses,
  address,
  abi,
  functionName,
  args,
  overrides,
  enabled
}: UseContractProps<CustomContractArgs>): string | undefined {
  const [contractInputData, setContractInputData] = useState<string | undefined>();
  const { data: signer } = useSigner();
  const { chainGroup } = useChain();
  const contract = useMemo(() => {
    if (!signer) {
      return null;
    }
    return new ethers.Contract(address ? address : getAddressByGroup(addresses, chainGroup), abi, signer);
  }, [addresses, address, abi, chainGroup, signer]);

  useEffect(() => {
    if (contract && enabled) {
      buildContractCall<CustomContractArgs>(contract, functionName, args, overrides).then(
        (inputData: string | undefined) => {
          if (inputData) {
            setContractInputData(inputData);
          }
        }
      );
    }
  }, [contract, functionName, args, overrides, enabled]);

  return contractInputData;
}
```

**Comment by jxom** [comment]
So I can assist further, what are you doing with `contractInputData`? Where is that being used?

**Reply by onchainguy-btc** [reply]
@jxom I'm using the `contractInputData` here: 

```typescript
const dataToSign = useMemo(() => {
    return {
      address: evmAddress,
      commodity: commodityForSignature,
      commodity_amount: commodityAmount,
      pk_id: 'key1',
      sc_address: contractAddress,
      sc_id: uuidv4(),
      sc_input_data: contractInputData
    };
  }, [contractAddress, commodityAmount, commodityForSignature, contractInputData, evmAddress]);
```

So I'm putting it into an object which is later JSON encoded and then sent to a server where it is signed. This is needed for allowing credit card payments from our frontend. The credit card provider needs the raw hex decoded input data to work.

**Comment by cxss** [comment]
I also create a contract instance like this:
const signer = useSigner();
const contract = new ethers.Contract(address, abi, signer);
it works in older version which has "useSigner".
In new version, you can get a signer by following code:
function useSigner({ chainId } = {}) {
  const { data: walletClient } = useWalletClient({ chainId })
  return React.useMemo(
    () => (walletClient ? walletClientToSigner(walletClient) : undefined),
    [walletClient],
  )
}
please refer to https://wagmi.sh/react/ethers-adapters

**Comment by cxss** [comment]
I also create a contract instance like this:
const signer = useSigner();
const contract = new ethers.Contract(address, abi, signer);
it works in older version which has "useSigner".

In new version, I get a signer by following code:
function useSigner({ chainId } = {}) {
  const { data: walletClient } = useWalletClient({ chainId })
  return React.useMemo(
    () => (walletClient ? walletClientToSigner(walletClient) : undefined),
    [walletClient],
  )
}
please refer to https://wagmi.sh/react/ethers-adapters
